db/src/db.rs

```diff
@@ -827,6 +855,87 @@ pub fn update_partition_map(conn: &rusqlite::Connection, partition_map: &Partiti
         .chain_err(|| "Could not update partition map")
 }
 
+/// Update the metadata materialized views based on the given metadata report.
+///
+/// This updates the "entids", "idents", and "schema" materialized views, copying directly from the
+/// "datoms" and "transactions" table as appropriate.
+pub fn update_metadata(conn: &rusqlite::Connection, _old_schema: &Schema, new_schema: &Schema, metadata_report: &metadata::MetadataReport) -> Result<()>
+{
+    use metadata::AttributeAlteration::*;
+
+    // Populate the materialized view directly from datoms (and, potentially in the future,
+    // transactions).  This might generalize nicely as we expand the set of materialized views.
+    // TODO: consider doing this in fewer SQLite execute() invocations.
+    // TODO: use concat! to avoid creating String instances.
+    if !metadata_report.idents_altered.is_empty() {
+        // Idents is the materialized view of the [entid :db/ident ident _ added] slice of
+        // transactions.  This keeps outdated idents around, just like Datomic does.
+        conn.execute(format!("DELETE FROM idents").as_str(),
+                     &[])?;
+        // Take the last, as determined by tx, added [e :db/ident v _ true].  This will pick up
+        // :db/idents that have been retracted and/or re-purposed.
+        conn.execute(format!("INSERT INTO idents SELECT e, a, v, value_type_tag FROM datoms WHERE a IN {}", entids::IDENTS_SET.as_str()).as_str(),
```

Comment:
I expected this to be efficient because there's an index on `e a v value_type_tag`.  Do you think it would be faster to have a bunch of `WHERE a = ? OR a = ?` clauses?  A little playing with SQLite suggests the query plans are the same:
```
sqlite> EXPLAIN QUERY PLAN INSERT INTO idents SELECT e, a FROM datoms WHERE a = 0 OR a = 1;
0|0|0|SEARCH TABLE datoms USING COVERING INDEX idx_datoms_aevt (a=?)
0|0|0|EXECUTE LIST SUBQUERY 1
sqlite> EXPLAIN QUERY PLAN INSERT INTO idents SELECT e, a FROM datoms WHERE a IN (0, 1);
0|0|0|SEARCH TABLE datoms USING COVERING INDEX idx_datoms_aevt (a=?)
0|0|0|EXECUTE LIST SUBQUERY 1
```
Feel free to file follow-up if you're not convinced.